Adds a aws_flow_log resource with unit and integration testing.

miah opened the pull request:
The aws_flow_log resource can search by subnet, vpc, or flow log ID.
It can match on name, resource_id, and log_group_name

Fixes #2581

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Adds a aws_flow_log resource with unit and integration testing.

The aws_flow_log resource can search by subnet, vpc, or flow log ID.
It can match on name, resource_id, and log_group_name

Fixes #2581

Signed-off-by: Miah Johnson <miah@chia-pet.org>

clintoncwolfe commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,82 @@
+platform: aws
+---
+
+# aws\_flow\_low
```
s/low/log/

clintoncwolfe commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,82 @@
+      it { should exist }
+    end
+
+## Resource Parameteres
```
s/Parameteres/Parameters/

clintoncwolfe commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,82 @@
+## Resource Parameteres
+### flow_log_id
+
+This resource expects a single parameter that uniquely identifies the Flow Log. You may pass it as a string, or as the value in a hash:
```
Because you can also search by subnet or VPC ID, I think we might want to soften "expects a single parameter" to something like "accepts a single parameter or other search terms"

clintoncwolfe commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,82 @@
+      it { should exist }
+    end
+
+## Matchers
```
Please add a 'Properties' section, and move name, log_group_name, and resource_id to that section.

clintoncwolfe commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,82 @@
+
+For a full list of available matchers, please visit our [matchers page](https://www.inspec.io/docs/reference/matchers/).
+
+### name
```
I think we should refer to this as `flow_log_id`, not as name.  I didn't see anything about a name attribute in the describe_flow_logs API docs.  Don't want to introduce confusion :)

clintoncwolfe commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,82 @@
+  describe aws_flow_log('fl-9c718cf5') do
+    its('resource_id') { should cmp 'subnet-c6a4319c' }
+  end
+
```
It might be useful to have a mid-level property, `resource_type`, that returns subnet, vpc, or eni. Or some matchers, attached_to_vpc?, attached_to_subnet?, etc.

clintoncwolfe commented on lib/resources/aws/aws_flow_log.rb:
```diff
@@ -0,0 +1,69 @@
+  name 'aws_flow_log'
+  supports platform: 'aws'
+  desc ''
+  example ''
```
Probably should put something here.

clintoncwolfe commented on lib/resources/aws/aws_flow_log.rb:
```diff
@@ -0,0 +1,69 @@
+
+    if validated_params.empty?
+      raise ArgumentError,
+            'aws_flow_log requires a parameter: flow_log_id, subnet_id, vpc_id'
```
Might want to add the word 'or' in there; otherwise people might think you have to supply all three.

clintoncwolfe commented on lib/resources/aws/aws_flow_log.rb:
```diff
@@ -0,0 +1,69 @@
+      raise ArgumentError,
+            'aws_flow_log requires a parameter: flow_log_id, subnet_id, vpc_id'
+    end
+
```
Should we do regexes on the subnet, flow log, or VPC id?

clintoncwolfe commented on test/integration/aws/default/build/flow_log.tf:
```diff
@@ -0,0 +1,68 @@
+output "fl-subnet" {
+  value = "${aws_flow_log.fl-subnet.id}"
+}
+resource "aws_cloudwatch_log_group" "test_log_group" {
```
Could we give this a different name, perhaps something specific to flow logs?

clintoncwolfe commented on test/integration/aws/default/build/flow_log.tf:
```diff
@@ -0,0 +1,68 @@
+  traffic_type   = "ALL"
+}
+
+output "fl-vpc" {
```
Our fixture naming standard is <resource_type>_<fixture_name>_<property>.  So you might name this `flow_log_alpha_vpc_id` or `flow_log_vpc_vpc_id`.

clintoncwolfe commented on test/integration/aws/default/build/flow_log.tf:
```diff
@@ -0,0 +1,68 @@
+resource "aws_flow_log" "fl-vpc" {
```
You might consider naming these more generically - alpha, beta, gamma, etc.  This make it easier to write tests later that are not interested in what the flow log is watching.  Ping me if that doesn't make sense :)

clintoncwolfe commented on test/integration/aws/default/build/flow_log.tf:
```diff
@@ -0,0 +1,68 @@
+  name = "test_log_group"
+}
+
+resource "aws_iam_role" "test_role" {
```
Same - could we get a name that is specific to flow log testing?

Terraform's namespace is flat, so if anyone ever uses that name, we have a collision.

clintoncwolfe commented on test/integration/aws/default/build/flow_log.tf:
```diff
@@ -0,0 +1,68 @@
+resource "aws_flow_log" "fl-subnet" {
+  log_group_name = "${aws_cloudwatch_log_group.test_log_group.name}"
+  iam_role_arn   = "${aws_iam_role.test_role.arn}"
+  subnet_id      = "${aws_subnet.subnet_01.id}"
```
It looks like aws_subnet.subnet_01 is defined in another file.  #2915 will require us to avoid doing that, because we will need to keep the services isolated.  You could look up the default subnet on the default VPC, or make your own subnet here.

clintoncwolfe commented on test/integration/aws/default/verify/controls/aws_flow_log.rb:
```diff
@@ -0,0 +1,44 @@
+fixtures = {}
+%w[fl-vpc fl-subnet subnet_01_id ec2_security_group_default_vpc_id].each do |fixture_name|
```
If you need ``ec2_security_group_default_vpc_id`, you should export it from the flow-logs.tf, and give it a flow-logs specific name.

clintoncwolfe commented on test/integration/aws/default/verify/controls/aws_flow_log.rb:
```diff
@@ -0,0 +1,44 @@
+  )
+end
+
+control 'aws_flow_log flow_log_id exists' do
```
I know the naming stuff can seem pedantic, but without it, blocks like this can be really confusing.  Below, only the exists line is clear; everything else requires (for me, at least) a bit of mental gymnastics to understand what is being tested against what.  I'm dumb, don't make me think!

clintoncwolfe commented on test/unit/resources/aws_flow_log_test.rb:
```diff
@@ -0,0 +1,74 @@
+
+  def test_search_miss
+    refute AwsFlowLog.new(flow_log_id: 'fl-12341234').exists?
+    refute AwsFlowLog.new(flow_log_id: 'fl-12341234').log_group_name
```
Might use `assert_nil` here - IIRC, they are going to change this behavior in the next version of MiniTest. Besides, you're checking for nilly-ness, not falsiness, right?

clintoncwolfe commented on test/unit/resources/aws_flow_log_test.rb:
```diff
@@ -0,0 +1,74 @@
+
+  class Basic < AwsBackendBase
+    def describe_flow_logs(query)
+      resp = Aws::EC2::Types::DescribeFlowLogsResult.new(
```
:100: for using the SDK's typing

clintoncwolfe reviewed (changes requested):
Great work @miah!  I just had a few naming requests - mostly internal, in the integration tests.  The implementation looks great, and aside from a typo or two, so do the docs.  I also suggest a derived property or 3 around the associated resource type, but that can wait for a later PR (or never).

miah committed:
Fixed typos and wording in documentation.
Renamed `name` property to `flow_log_id`
Use assert_nil rather than refute, we are testing for nil values not truthiness
Reworded control description
Add forgotten example/description in resource
Rename all flow log terraform resources to be unique
Add flow log subnet in terraform
Only export terraform flow log values declared by flow log tf.
Cleaned up some trailing spaces in TESTING doc.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Add matchers for `be_attached_to_eni`, `be_attached_to_subnet`, and `be_attached_to_vpc`.
Add mid-level property for `resource_type` where it returns the type of
resource the flow log is attached to.
Add tests for matchers and property.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Document new property and matchers.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

clintoncwolfe commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+    end
+
+## Properties
+### name
```
Please change to `flow_log_id`.  There is no `name` property.

clintoncwolfe commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+### resource_type
+
+The `resource_type` property tests the type of resource the Flow Log is attached to.
+The property will return `eni`, `subnet`, and `vpc`.
```
I think 'or' makes more sense than 'and' here - it will only return one of them.

clintoncwolfe commented on lib/resources/aws/aws_flow_log.rb:
```diff
@@ -0,0 +1,95 @@
+  include AwsSingularResourceMixin
+
+  def to_s
+    "AWS Flow Log #{@flow_log_id}"
```
If someone searches for a flow log by vpc_id or subnet_id, and it does not exist, @flow_log_id will be nil.  Could we add some logic to improve to_s in that case?

clintoncwolfe reviewed (changes requested):
Two small docs changes and one UX change for to_s - looks very good overall!

miah committed:
Add tests for #to_s

Always return a valid id for to_s.
If you search by subnet_id, or vpc_id, and the flow log is not found
to_s will show the id you searched with. eg 'AWS Flow Log subnet-12341234'

Fixed docs

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Merge branch 'master' into miah/2581

clintoncwolfe reviewed (approved):
Looks great - thanks @miah!

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+    end
+
+## Resource Parameters
+### flow_log_id
```
Need `\_` here since it is a heading.

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+      it { should exist }
+    end
+
+### subnet_id
```
Need `\_` here since it's a heading.

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+      it { should exist }
+    end
+
+### vpc_id
```
Need `\_` here since it's a heading.

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+    end
+
+## Properties
+### flow_log_id
```
Need `\_` here since it's a heading.

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+    its('flow_log_id') { should cmp 'fl-9c718cf5' }
+  end
+
+### log_group_name
```
Need `\_` here since it's a heading.

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+### resource_id
+
+The `resource_id` property tests the id of the associated VPC, subnet, or network interface.
+  describe aws_flow_log('fl-9c718cf5') do
```
Not sure if a new line is needed above here, but it would match the others.

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+    its('resource_id') { should cmp 'subnet-c6a4319c' }
+  end
+
+### resource_type
```
Need `\_` here since it's a heading.

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+      it { should_not exist }
+    end
+
+### be_attached_to_eni
```
Need `\_` here since it's a heading.

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+      it { should be_attached_to_eni }
+    end
+
+### be_attached_to_subnet
```
Need `\_` here since it's a heading.

jerryaldrichiii commented on docs/resources/aws_flow_log.md.erb:
```diff
@@ -0,0 +1,117 @@
+      it { should be_attached_to_subnet }
+    end
+
+### be_attached_to_vpc
```
Need `\_` here since it's a heading.

jerryaldrichiii commented on lib/resources/aws/aws_flow_log.rb:
```diff
@@ -0,0 +1,101 @@
+  example "
+    describe aws_flow_log('fl-9c718cf5') do
+      it { should exist }
+    end"
```
I usually see another newline after `end`, not sure if it matters though.

jerryaldrichiii commented on test/unit/resources/aws_flow_log_test.rb:
```diff
@@ -0,0 +1,130 @@
+  def test_rejects_unrecognized_params
+    assert_raises(ArgumentError) { AwsFlowLog.new(shoe_size: 9) }
+  end
+
```
Extra new line.

jerryaldrichiii reviewed (changes requested):
Great work @miah! Only found some style related things.

clintoncwolfe committed:
Feedback from Jerry

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

jerryaldrichiii reviewed (approved):
Thank you @clintoncwolfe for sparing @miah from my pedantry.
